soc/st/stm32/common/stm32_iocell.c

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2024 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/types.h>
+#include <soc.h>
+#include <zephyr/devicetree.h>
+
+#include <stm32_ll_system.h>
+#include <stm32_ll_pwr.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(iocell, CONFIG_SOC_LOG_LEVEL);
+
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+#include <zephyr/dt-bindings/power/stm32h7rs_iocell.h>
+#elif defined(CONFIG_SOC_SERIES_STM32N6X)
+#include <zephyr/dt-bindings/power/stm32n6_iocell.h>
+#define STM32N6_IOCELL_FIXEDVALUE (0x00000287)
+#endif
+
+#define DT_DRV_COMPAT     st_stm32_iocell
+#define STM32_IOCELL_NODE DT_NODELABEL(iocell)
+
+/* Returns 1 if HSLV can be confgured */
+static inline int iocell_get_hslv_safeguard(uint16_t domain)
+{
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	if (domain == STM32_DT_IOCELL_VDDIO) {
+		return (FLASH->OBW1SR & FLASH_OBW1SR_VDDIO_HSLV) == FLASH_OBW1SR_VDDIO_HSLV;
+	} else if (domain == STM32_DT_IOCELL_XSPI1) {
+		return (FLASH->OBW1SR & FLASH_OBW1SR_XSPI1_HSLV) == FLASH_OBW1SR_XSPI1_HSLV;
+	} else if (domain == STM32_DT_IOCELL_XSPI2) {
+		return (FLASH->OBW1SR & FLASH_OBW1SR_XSPI2_HSLV) == FLASH_OBW1SR_XSPI2_HSLV;
+	} else {
+		return -1;
+	}
+#elif defined(CONFIG_SOC_SERIES_STM32N6X)
+	static uint32_t hconf1 = 0xFFFFFFFF;
+	uint32_t shift = 16 - domain;
+
+	if (hconf1 == 0xFFFFFFFF) {
+		WRITE_REG(BSEC->OTPCR, 124);
```

Comment:
```suggestion
		stm32_reg_write(&BSEC->OTPCR, 124);
```
See 2a52a2b7ff48f0deeadcd3aece7f9064a54b130a.